File: backend/utils/cache.py
from datetime import datetime
from typing import Dict, Any

class Cache:
    def __init__(self, expiry=3600):
        self._cache: Dict[str, Dict[str, Any]] = {}
        self.expiry = expiry
    
    def get(self, key: str) -> Any:
        if key in self._cache:
            entry = self._cache[key]
            now = datetime.now().timestamp()
            if now - entry["timestamp"] < self.expiry:
                return entry["data"]
        return None
    
    def set(self, key: str, value: Any) -> None:
        self._cache[key] = {
            "timestamp": datetime.now().timestamp(),
            "data": value
        }
    
    def clear(self) -> None:
        self._cache.clear()

cache = Cache()

File: backend/src/utils/cache.ts
type CacheEntry<T> = {
  value: T;
  expires: number;
};

class SimpleCache<T = any> {
  private store = new Map<string, CacheEntry<T>>();

  constructor(private ttlMs: number = 3600 * 1000) {
    this.startCleanup();
  }

  get(key: string): T | undefined {
    const cached = this.store.get(key);
    if (!cached) return undefined;
    if (Date.now() > cached.expires) {
      this.store.delete(key);
      return undefined;
    }
    return cached.value;
  }

  set(key: string, value: T, ttlMs: number = this.ttlMs) {
    this.store.set(key, { value, expires: Date.now() + ttlMs });
  }

  clear() {
    this.store.clear();
  }

  private startCleanup(interval: number = 300000 /* 5 minutes */) {
    setInterval(() => this.cleanupExpired(), interval);
  }

  private cleanupExpired() {
    const now = Date.now();
    for (const [key, entry] of this.store.entries()) {
      if (now > entry.expires) {
        this.store.delete(key);
      }
    }
  }
}

export const cache = new SimpleCache();
